src/main/scala/org/virtuslab/internal/load/reader/ReaderCtx.scala

```diff
@@ -36,20 +36,17 @@ case class ReaderCtx(
         closeOpenedCollectionMapping(indent)
       case _ => ()
 
-  def appendSequence(indent: Int): Token =
-    stateStack.push(ReaderState.Sequence(indent))
-    Token.SequenceStart
+  def appendState(state: ReaderState): Unit = stateStack.push(state)
 
-  def appendMapping(indent: Int): Token =
-    stateStack.push(ReaderState.Mapping(indent))
-    Token.MappingStart
+  def closeOpenningFlowMapping(): List[Token] =
+    stateStack.pop() match
+      case Some(ReaderState.FlowMapping) => List(Token.FlowMappingEnd)
+      case _                             => Nil
 
-  def popTokenFromStack: Token =
+  def closeOpenningSequence(): List[Token] =
     stateStack.pop() match
-      case Some(ReaderState.Sequence(_)) => Token.SequenceEnd
-      case Some(ReaderState.Mapping(_))  => Token.MappingEnd
-      case Some(ReaderState.Document)    => popTokenFromStack
-      case _                             => Token.StreamEnd
+      case Some(ReaderState.Sequence(_)) => List(Token.SequenceEnd)
+      case _                             => Nil
```

Comment:
It seems that `closeX` and `stateStack.pop()` could use some error handling in future iterations :D

Reply:
Good point, it should return error from input tokenising